### DISCUSSION TITLE
Azure communications services for email producing 504 when deployed by Vercel

### DISCUSSION THREAD

**Discussion by ghost** [vercel/community#1095] [Help] [answered]
I am using Azure communications services to send emails. I have set it all up in the cloud and locally and emails are working quite nicely. When I deploy to vercel, my nextjs api route fails with a 504. 

```
const connectionString =
    process.env.AZURE_COMMUNICATIONS_CONNECTION_STRING || "";
  var client = new EmailClient(connectionString);
  const emailMsg: EmailMessage = {
    sender: "public@mydomain.com",
    content: {
      subject: "Confirm your email",
      plainText: `
        Hello and thank you for registering
    `
    },
    recipients: {
      to: [
        {
          email,
        },
      ],
    },
  };

await client.send(emailMsg);
res.json({});
```

The await is causing the timeout failure. I understand vercel functions should take less then 5 seconds. But Azure is quite reliable and I cannot imagine their client to be taking more than 10 seconds unless there is some type of connectivity issue between Vercel and Azure.

**Comment by swarnava** [comment, marked as answer]
Hi,

You need to ensure the API response back within 10 seconds. Here are couple of tips: https://vercel.com/guides/what-can-i-do-about-vercel-serverless-functions-timing-out

The best way to see how long your Function is taking for a particular task is by logging its execution time. The best way to do that is by using `console.time()`. Here’s an example:

```js
{
  console.time('doStuff')
  const data = await doStuff()
  console.timeEnd('doStuff')
  return {
    body: JSON.stringify(data)
    statusCode: 200
  }
```

In your Function console, you’d see something like: `doStuff: # seconds`. This will show how long your `doStuff()` call took. Again, depending on your Function code - for instance if it is using features like `async/await` statements, Promises, etc. you might have to adapt the above to correctly calculate those timings.

Once you know what piece of code is acting as a bottleneck in your Function, you can work on changing that.

**Reply by ghost** [reply]
Yes I already knew this. The await is taking more than 10 seconds. The answer is why? It seems like there are connectivity problems between Vercel edge and Azure Communications.

**Reply by amyegan** [reply]
I think there were some problems related to Azure over the weekend. It was a major holiday shopping weekend in the U.S. so many platforms and services were under a lot of stress. 😵‍💫

Are you still running into this time out problem? And, if you are, have you tried Swarnava's suggestion to get more timing details?

**Reply by ghost** [reply]
Good call! It works just fine now. Must have been all that exciting shopping. Thank you both for jumping in and giving me a sanity point that there really was a traffic flow issue between Vercel and Azure. Now I am able to remove my Azure FaaS and directly call the Azure mailer API from Vercel. Sweet!
